query-engine/query-engine/src/server/mod.rs

```diff
@@ -241,25 +242,42 @@ fn playground_handler() -> Response<Body> {
 }
 
 async fn handle_metrics(state: State, req: Request<Body>) -> Result<Response<Body>, hyper::Error> {
-    if let Some(query) = req.uri().query() {
+    let format = if let Some(query) = req.uri().query() {
         if query.contains("format=json") {
-            let metrics = state.cx.metrics.to_json(Default::default());
+            MetricFormat::Json
+        } else {
+            MetricFormat::Prometheus
+        }
+    } else {
+        MetricFormat::Prometheus
+    };
 
-            let res = Response::builder()
-                .status(StatusCode::OK)
-                .header(CONTENT_TYPE, "application/json")
-                .body(Body::from(metrics.to_string()))
-                .unwrap();
+    let body_start = req.into_body();
+    // block and buffer request until the request has completed
+    let full_body = hyper::body::to_bytes(body_start).await?;
 
-            return Ok(res);
-        }
+    let global_labels: HashMap<String, String> = match serde_json::from_slice(full_body.as_ref()) {
+        Ok(map) => map,
+        Err(_e) => HashMap::new(),
+    };
+
+    if format == MetricFormat::Json {
+        let metrics = state.cx.metrics.to_json(global_labels);
+
+        let res = Response::builder()
+            .status(StatusCode::OK)
+            .header(CONTENT_TYPE, "application/json")
+            .body(Body::from(metrics.to_string()))
+            .unwrap();
+
+        return Ok(res);
     }
 
-    let metrics = state.cx.metrics.to_prometheus(Default::default());
+    let metrics = state.cx.metrics.to_prometheus(global_labels);
 
     let res = Response::builder()
         .status(StatusCode::OK)
-        .header(CONTENT_TYPE, "application/json")
+        .header(CONTENT_TYPE, "application/text")
```

Comment:
Prometheus spec tells me it should be `text/plain; version=0.0.4`? https://prometheus.io/docs/instrumenting/exposition_formats/#text-based-format